app/src/main/java/com/fstyle/structure_android/utils/Validator.java

```diff
@@ -0,0 +1,359 @@
+        }
+        mContext = context;
+        mValidatedMethods = cacheValidatedMethod();
+        mAllErrorMessage = getAllErrorMessage(clzz.getClass());
+    }
+
+    private SparseArray<Method> cacheValidatedMethod() {
+        SparseArray<Method> methods = new SparseArray<>();
+        for (Method method : this.getClass().getDeclaredMethods()) {
+            if (method.isAnnotationPresent(ValidMethod.class)) {
+                ValidMethod validMethod = method.getAnnotation(ValidMethod.class);
+                for (int type : validMethod.type()) {
+                    methods.put(type, method);
+                }
+            }
+        }
+        return methods;
+    }
+
+    private SparseArray<Integer> getAllErrorMessage(Class clzz) {
+        SparseArray<Integer> errorMessages = new SparseArray<>();
+        for (Field field : clzz.getDeclaredFields()) {
+            Validation annotation = field.getAnnotation(Validation.class);
+            if (annotation == null) {
+                continue;
+            }
+            Rule[] rules = annotation.value();
+            field.setAccessible(true);
+            for (Rule rule : rules) {
+                for (int type : rule.types()) {
+                    errorMessages.put(type, rule.message());
+                }
+            }
+        }
+        return errorMessages;
+    }
+
+    /**
+     * Check if an object in Validator Model with annotation {@link Optional} is valid
+     */
+    private boolean isValidOptional(Object factor) {
+        if (factor instanceof Integer) {
+            return (Integer) factor == 0;
+        }
+        if (factor instanceof Long) {
+            return (Long) factor == 0;
+        }
+        if (factor instanceof Float) {
+            return (Float) factor == 0.0f;
+        }
+        if (factor instanceof Double) {
+            return (Double) factor == 0.0f;
+        }
+        if (factor instanceof String) {
+            return TextUtils.isEmpty((String) factor);
+        }
+        return (factor instanceof Boolean && !((Boolean) factor)) || factor == null;
+    }
+
+    /**
+     * Validate all field.
+     */
+    private boolean validate(Object factor, Rule[] rules, boolean isOptional) {
+        boolean isValid = true;
+        for (Rule rule : rules) {
+            for (int type : rule.types()) {
+                Method method = mValidatedMethods.get(type);
+                if (method == null) {
+                    continue;
+                }
+                try {
+                    boolean valid = (isOptional && isValidOptional(factor)) || TextUtils.isEmpty(
+                            (String) method.invoke(this, factor));
+                    if (!valid) {
+                        isValid = false;
+                    }
+                } catch (IllegalAccessException e) {
+                    isValid = false;
+                    Log.e(TAG, "validate: ", e);
```

Comment:
WARNING: [AvoidDuplicateLiterals] The String literal "validate: " appears 4 times in this file; the first occurrence is on line 155
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals